src/components/Upload/FileUploader.js

```diff
@@ -52,23 +86,63 @@ export default class FileUploader extends Component {
     return result;
   }
 
+  convertIPUZ(readerResult) {
+    const {grid, info, circles, shades, across, down} = iPUZtoJSON(readerResult);
+
+    const result = {
+      grid,
+      circles,
+      shades,
+      info,
+      clues: {across, down},
+    };
+
+    return result;
+  }
+
+  attemptPuzzleConversion(readerResult, fileType) {
+    if (fileType === 'puz') {
+      return this.convertPUZ(readerResult);
+    } else if (fileType === 'ipuz') {
+      return this.convertIPUZ(readerResult);
+    } else if (fileType === 'jpz') {
+      throw new UnsupportedFileTypeError(fileType);
+    } else {
+      const guessedFileType = fileTypeGuesser(readerResult);
+      if (!guessedFileType) {
+        throw new UnknownFileTypeError(fileType);
+      } else {
+        return this.attemptPuzzleConversion(readerResult, guessedFileType);
+      }
+    }
+  }
+
   onDrop(acceptedFiles) {
     const file = acceptedFiles[0];
+    const fileType = file.name.split('.').pop();
     const reader = new FileReader();
     const {success, fail} = this.props;
     reader.addEventListener('loadend', () => {
       try {
-        const puzzle = this.convertPUZ(reader.result);
+        const puzzle = this.attemptPuzzleConversion(reader.result, fileType);
         if (this.validPuzzle(puzzle)) {
           success(puzzle);
         } else {
           fail();
         }
       } catch (e) {
+        let defaultTitle = 'Something went wrong';
+        let defaultText = `The error message was: ${e.message}`;
+        let defaultIcon = 'warning';
+
+        if (e?.errorTitle) defaultTitle = e.errorTitle;
+        if (e?.errorText) defaultText = e.errorText;
+        if (e?.errorIcon) defaultIcon = e.errorIcon;
+
         swal({
-          title: `Invalid .puz file`,
-          text: `The uploaded file is not a valid .puz file.`,
-          icon: 'warning',
+          title: defaultTitle,
+          text: defaultText,
+          icon: defaultIcon,
```

Comment:
Abstracted the error handling out a bit to show some of the messages of the errors:
![error-handling](https://user-images.githubusercontent.com/463175/230793192-134a2368-0b5d-420d-a099-1ac9d18615d2.gif)

Reply:
i'd usually do something like this, or even just inline it into the call to `swal`, but fine with this way too.
```
        const defaultTitle = e?.errorTitle ?? 'Something went wrong';
```